### ISSUE TITLE
Updating a Model with Unique field

### ISSUE THREAD

**Issue by raeldc** [jonathangeiger/kohana-jelly#16]
Just found a bug where I'm not allowed to update an object if a unique field was changed but wasn't really changed.

Here's the code:

```
//this item has a unique field 'title' with value 'Item Title'
$item = Jelly::factory('item')->load($id);

$values = array(
    'title' = 'Item Title', //this is exactly the same title.
);

$item->set($values);
try 
{
    $item->save();
    $this->request->response = 'Item Saved!'
} 
catch (Validate_Exception $e) 
{
    Kohana::debug($e->array->errors());
}
```

Error message says title must be unique. This situation happens when I use `$item->set($_POST)`

Thanks!

**Comment by banks** [comment]
Looks like we need to do an equality test when setting to determine if the value is actually different before adding to 'changed' - I thought this happened already but clearly it doesn't.

**Comment by raeldc** [comment]
Is the fix as simple as this in line 384 of `Jelly_Core_Model`?

```
        else if ($alias === FALSE && $field = $this->field($key))
        {
            // Check if the original value has really changed
            if ($this->_original[$field->name] != $value) 
            {
                $data_location[$field->name] = $field->set($value);
            }
            // Invalidate the cache
            if (array_key_exists($field->name, $this->_retrieved))
            {
                unset($this->_retrieved[$field->name]);
            }
        }
```

Or there are some things that must be considered?

**Comment by banks** [comment]
That looks right to me. If I get a chance I'll put this in - Jon may well get there first though!

**Comment by jonathangeiger** [comment]
Closed by 33c12846b87597324fbd3f563a60a96fcdcbb7d7. Thanks raeldc!

**Comment by raeldc** [comment]
Thanks Jonathan, but test further, coz at first I tried 

`if ($this->_original[$field->name] !== $value)`

But I had problems with numeric characters. `$this->_original` already has correct types for each field like `(int)` but when you pass a $_POST, some fields usually are numeric `string`. So using `!==` will still mark fields as "changed" because of strict type comparison.

I suggest you simply use value comparison:

`if ($this->_original[$field->name] != $value)`

**Comment by jonathangeiger** [comment]
Alright. Using != breaks some others things. I'll look into it more.

**Comment by jonathangeiger** [comment]
Closed by d22dff1e9dc2b484e4ca0d735deddb12a66c18d2. Fields that are in_db are not changed if they haven't really changed. Thanks again for the help raeldc!

**Comment by jonathangeiger** [comment]
I'm going to re-open this because I'm not sure it's entirely fixed. I'm kind of stumped about how to _fully_ fix it (though it is rather late now, and I'm a bit groggy).

Here's the issue, in a nutshell:

```
Jelly::factory('model')
    ->set($_POST)
    ->save($id);
```

Now, say $_POST contains a field that is declared in the model to be unique. Let's also say that the user didn't change that value when submitting the form. The original fix for this issue was to compare the value in the model with the incoming value, the problem is that, in this case, the model isn't loaded yet, so the comparison will allow the value to be counted as "changed" data.

When it comes time to save() the model (notice we're passing an ID so it's updating), the field is validated because it was counted as changed. The validation fails because the value exists in the database.

So I'm not entirely sure what the solution to this is. The model can't be passed to the validation callback, so we can't, for example, compare the two models to see if they're actually the same model.

Any ideas?

**Comment by raeldc** [comment]
Pardon my simplistic answer. But my obvious solution is to auto load the model when `->save($id)` is used. That's what I've always been doing, that's why I didn't know `->save($id)` is even possible. I've always been doing this:

```
$model = Jelly::select('model')->load($id);

if ($model->loaded())
{
    try{
        $model->set($_POST)->save();
    }
    catch(Validate_Exception $e)
    {
        //debug here
    }
}else
{
    //error saying the item you're trying to edit is non-existent
}
```

I guess in the `save()` method, you can do something like this:
1. Detect if a key is passed this means that it's trying to update.
2. Load the model based on the key.
3. Since `set($_POST)` is called first then data is in `this->_changed` so set it again. 
   
    $data = $this->_changed;
   $this->set($data);
4. Proceed with validation, then on to saving.

This is obviously a very simplistic solution so I am guessing you already came across this but ignored it because of other issues it might cause?

**Comment by jonathangeiger** [comment]
The only problem with that is that the point of passing an $id to save() is so that you don't have to issue the extra query to actually find the record.

**Comment by raeldc** [comment]
I don't know what's wrong with getting an extra query. I'm pretty sure it's not a performance concern, or is it? It's just the simplest or even the only solution I can think of. If someone can think of a nicer solution without an extra query, that guy must be a genius.

**Comment by banks** [comment]
> I don't know what's wrong with getting an extra query. I'm pretty sure it's not a performance concern, or is it?

Yes in that there is no point in having save($key) at all if it doesn't save a query. If we have to change it so that this loads anyway then we may as well remove it and keep the API more clean and consistent.

**Comment by raeldc** [comment]
I thought the `save($key)` was just a shortcut. Like I said, I never even knew I can do that. But I don't see how it's inconsistent with the API to do the checking of values from the database to see what changed or not. It's an ORM after all and not a simple database abstraction layer.

But then again it's your decision. I know you guys are good enough to come up with the best solution :).

**Comment by jonathangeiger** [comment]
> I don't know what's wrong with getting an extra query. I'm pretty sure it's not a performance concern, or is it? It's just the simplest or even the only solution I can think of.

I don't have any hard data as to whether it is or not. I just know that it's a feature that got a lot of attention in Kohana's ORM. It might be a case of people just trying to micro-optimize though.

> But I don't see how it's inconsistent with the API to do the checking of values from the database to see what changed or not.

I think Banks and I just try to make sure there's only 1 canonical way to do things. Otherwise you have multiple points of failure and more to test. I'm not saying we've done that perfectly with this release, but if you can do this:

```
Jelly::select('model', $id)
    ->set($data)
    ->save();
```

Just as easily as you can do this:

```
Jelly::factory('model')
    ->set($data)
    ->save($id);
```

And they both do the exact same thing, why not just force people to always use the first way?

**Comment by glamorous** [comment]
Don't know if this is relevant but what should happen if you do this:

```
Jelly::factory('model')
    ->set($data)     // $data has a 'id' => $id_that_exists
    ->save();
```

Will this return an error or does Jelly the update instead of an insert?

**Comment by jjlegendre** [comment]
As an alternative, since the issue is only relevant with unique fields then can't you have the model load if there is a field in the model declared unique and not load otherwise.  That way it still avoids running the extra query unless it absolutely has to?  Just a suggestion.

**Comment by jonathangeiger** [comment]
> Don't know if this is relevant but what should happen if you do this:

An INSERT will be performed.

> As an alternative, since the issue is only relevant with unique fields then can't you have the model load if there is a field in the model declared unique and not load otherwise. 

This is probably what we'll have to end up doing. That or just saying you can't use unique fields with this feature. I'd definitely prefer the former though.

**Comment by glamorous** [comment]
@jonathan So the code will return an error for existing id?

**Comment by jonathangeiger** [comment]
Ah I see your question. Jelly doesn't perform any checks for primary keys at the moment, but most databases will throw an exception if you try to do that, so yes.

**Comment by banks** [comment]
Fixed in a branch pending Jon's approval. This is a fix for this issue that doen't resort to complex rules about when or not to perform the load query although it does use a slightly dubious solution. Jon, see what you make of [this](http://github.com/jonathangeiger/kohana-jelly/commit/8a2f75d92f820cee18bac81c33ba78a51a4d634b). It works and the tests prove it. 

The only slightly messy thing is that when lazy saving, a non-field key value is added to the validation data and an unnecessary filter has to be added in validate() to ensure the key value persists. This allows Field::_is_unique to filter out the actual row that's being updated from the check.

Although it' s not ideal I don't think it is a bad compromise for this situation. The other alternative - sometimes doing the extra query and sometimes not - seems just as 'bad' without the benefit of actually working as we want.

Anyway, if you feel it is a good enough solution, merge it to master and close this darned issue!

**Comment by jonathangeiger** [comment]
Merged to master in 8a2f75d92f820cee18bac81c33ba78a51a4d634b. Excellent work Paul. This is much better than any solution I thought of.
